### ISSUE TITLE
Feature interaction and correlation in EBM

### ISSUE THREAD

**Issue by britneyzeng** [interpretml/interpret#442]
It seems to me that EBM can only consider the interaction between two features. How can it perform well if the data contains interaction between more than two features? Have you checked the performance of EBM on data with more complex interactions?

Also, the website mentions that the boosting procedure is training one feature at a time in a round-robin fashion using a very low learning rate so that the feature order does not matter. What is the reason behind this? Wouldn't that become more biased toward certain features? And why round-robin cycling through features can mitigate the effects of co-linearity? I can see the original paper uses an experiment to show that, but not really get the reason behind this. If the order of learning those features is fixed in each iteration, how can that be the case?

**Comment by paulbkoch** [comment]
Hi @britneyzeng -- Actually, you can now build EBMs that consider as many dimensions as you like.  We haven't really had a chance to try 3 dimensional EBMs yet beyond simple testing, so this is an area worth exploring. Currently you need to specify the 3/4/5/etc way interactions manually, like this:

ebm=ExplainableBoostingClassifier(interactions=[(0,1,2), (4,5,6,9), ...])

Our automatic interaction detection doesn't support 3 way interactions yet, so you need to figure these out yourself.  There are a number of ways you could do this, but one way I might recommend is to first build an initial pairwise EBM, then extract a list of all the features used in the pairs and then just brute force include all the 3 ways that could be constructed from that set of features.

If I had to guess though, I would say that on most datasets 3-ways interactions will have limited value. I have two reasons for thinking this.  Firstly, for EBMs, pairs add surprisingly little benefit on many datasets already.  There are the odd datasets that seem to really need them, but for the most part this doesn't seem to be the case.  The bikeshare dataset is one that I know benefits from pairs, and I suspect might benefit further from 3-way interactions.

My second line of reasoning why triples are not likely to be extremely useful is because the gap in performance between EBMs and XGBoost is so small.  If you thought that triples were important and that XGBoost was capturing that benefit, then it's hard to explain why EBMs do so well on so many datasets given how similar XGBoost and EBMs are under the hood.


The cyclic aspect of EBMs is another thing that you can change now.  As of version 0.4.0 we support a "greediness" parameter.  This is set to 0 by default which means that we're still a cyclic algorithm by default.  If you set that parameter to 1, you'll get fully greedy boosting.  Setting it to values between 0 and 1 will make the algorithm more or less greedy in proportion to that value.  This is another feature that we haven't had a chance to really explore in full though.  Our benchmarking indicates though that values closer to 0.5 yield better performing models though.

Cyclic boosting doesn't solve the effects of co-linearity, but in our experience it does a fairly good job spreading credit between features that have co-linearity since the booster gives all features an equal number of boosting rounds.  With the learning rate set to the default of 0.01 a typical EBM typically requires somewhere in the range of 1000-3000 rounds of boosting per feature to reach early stopping.  If you had 100 features, this means a total of 100,000 to 300,000 rounds of boosting.  By the time you reach the 300,000th boosting step, the original order from the first step is mostly irrelevant.

**Comment by paulbkoch** [comment]
Hi @britneyzeng -- You might also be interested in this paper discussing purification.  EBMs without 3-way interactions will only be blind to the true purified 3-way interaction component.

https://arxiv.org/pdf/1911.04974.pdf
